GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m

```diff
@@ -323,17 +350,128 @@ - (void)removeEvents:(NSSet<NSNumber *> *)eventIDs {
   });
 }
 
-#pragma mark - GDTCORStorageProtocol
+- (nullable NSNumber *)
+    batchWithEventSelector:(nonnull GDTCORStorageEventSelector *)eventSelector
+           batchExpiration:(nonnull GDTCORClock *)expiration
+                onComplete:(nonnull void (^)(NSNumber *_Nullable batchID,
+                                             NSSet<GDTCOREvent *> *_Nullable events))onComplete {
+  NSNumber *batchID = [self nextBatchID];
+  if (!batchID) {
+    return nil;
+  }
+  dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+  __block NSSet<NSString *> *eventPaths;
+  [self pathsForTarget:eventSelector.selectedTarget
+              eventIDs:eventSelector.selectedEventIDs
+              qosTiers:eventSelector.selectedQosTiers
+            mappingIDs:eventSelector.selectedMappingIDs
+            onComplete:^(NSSet<NSString *> *_Nonnull paths) {
+              eventPaths = paths;
+              dispatch_semaphore_signal(sema);
+            }];
+  if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 10 * NSEC_PER_SEC)) != 0) {
```

Comment:
The call to pathsForTarget::::: also happens on the queue, so it needs to be outside the dispatch_async block below. Without the semaphore, the current thread will block anyway. The semaphore provides a timeout mechanism if there are too many operations queued on on the queue. 10 seconds just seemed like a reasonable number. I'm open to alternatives.

Reply:
Since `dispatch_async` calls can be safely nested even for a serial queue, I think, executing the remaining work can be done from the completion. We can try to deduce the callback hell if we split the implementation into separate methods. Though this approach won't provide the timeout feature if it is a requirement.